### ISSUE TITLE
TypeError: expected array of bytes

### ISSUE THREAD

**Issue by asenfter** [dask/fastparquet#455]
Hi, 
I am using partitioning mechanism when loading and writing my data. Writing works fine but when loading the data I run into some issues. Actually I do not know if this here is the right place to come up with this problem. Nevertheless, I hope for helpful hints.

All parquet files I load have the same columns. Unfortunately because of historical reasons the values of one particular column are stored in two different ways. But always consistent within one file. E.g. the column has float values in file A and stringified float values in file B.

``` 
# file A with float values in column
|  value  | partition |
+---------+-----------+
|  47.11  |     's'   |

# file B with stringified float values in column
|  value  | partition |
+---------+-----------+
| '47.11' |     'f'   |
```

When I load the files in one go and create a pandas DataFrame I have the problem that depending on the order of the files fed to my program it works or ends in an error: `TypeError: expected array of bytes`

* When I first load the file with float data followed by the file with string data no problem occurs.
* But when I first load the file with string data followed by the file with float data my program ends in a TypeError.

Is the mechanism wanted to work like this?

Actually I deal with a lot of data and converting all data to a common type (string or float) is not really an option. Is there any work-around which makes it possible to avoid the TypeError?

For better understanding I wrote a unittest which explains my problem by code. You can find it at https://github.com/asenfter/fp-test. It also contains a work-around but I can't tell if this solution is a good one.

**Comment by martindurant** [comment]
The parquet standard does expect that the dtypes are the same across all row-groups/files, although some limited "evolution" may be allowed (this is not implemented in fastparquet and tricky to encapsulate). 
In the case that the column is interpreted as "object" (because of text), it can contain any values, but the individual chunk is still being read as integers and converted - which is very inefficient.

In the usual case, data files are packages together into a directory to form a single data-set, and a special file, _metadata or _common_metadata is the authoritative arbiter or the schema.
In your case, you may want to consider the files independently and load one-by-one, using either pandas.concat or dask.dataframe.concat to construct the single logical data-set (possible with some .astype() calls before).

For your workaround, I am not sure I understand what the consequence is, without further digging.

**Comment by asenfter** [comment]
Thanks Martin for your answer. Ok I understand. I was just hoping that there is a out-of-the-box mechanism provided which solves my problem. 
So in your point of view the best and cleanest solution is to bring all data to a consistent format or loading the files sequentially and stacking the DataFrames after casting their type to String, right?
Although we do not understand the consequence of the work-around yet loading is still very inefficient because of the chunked conversion you mentioned above, right?

**Comment by martindurant** [comment]
Yes; although the "correct" dtype for your data seems to be int; which will be much more efficient if you want to do mathematical operations.
